DNS

Dirbaio opened the pull request:
Requires #573 

DNS in smoltcp! Overall design matches the rest of smoltcp, there should be no surprises :)

The DNS state machine is implemented as a DnsSocket, instead of using UdpSocket, because:
- It can send and receive UDP packets on the fly, avoiding the problems we ran into with DHCP (need big buffers, if they're full packets get dropped...).
- The UDP source port has to be randomized per query, so running N queries in parallel would require N UdpSockets.

`wire` is complete, `socket` works with a few missing things:

- [x] Remove hardcoded IP addresses in socket.
- [x] Support multiple servers, retry queries with other servers when one doesn't work.
- [x] Add timeouts
- [x] Allow specifying DNS names in human-friendly string format: `"rust-lang.org" -> b"\x09rust-lang\x03org\x00";`
- [x] Add protection against CNAME loops
- [x] Handle NXDOMAIN and other DNS server errors
- [x] Handle when query completes succesfully but there are no A/AAAA records.
- [ ] Add support for AAAA queries. -- will do later
- [ ] Perhaps add a way to query A+AAAA at the same time? -- will do later
- [x] Proper random DONE pending #547 
- [x] Fix ipv6-less build

birkenfeld commented on examples/dns.rs:
```diff
@@ -0,0 +1,94 @@
+    let name = &[
+        0x09, 0x72, 0x75, 0x73, 0x74, 0x2d, 0x6c, 0x61, 0x6e, 0x67, 0x03, 0x6f, 0x72, 0x67, 0x00,
+    ];
+    let name = &[
```
can't this be b"" strings to make the content a bit clearer?

crawford commented on src/socket/dns.rs:
```diff
@@ -0,0 +1,455 @@
+    /// The time-to-live (IPv4) or hop limit (IPv6) value used in outgoing packets.
+    hop_limit: Option<u8>,
+
+    // TEMPORARY rand implementation. TODO
```
Is the eventual idea to expose Rand as a public trait and add a type parameter to DnsSocket, allowing users to bring their own implementation (or fall back on this default implementation)? Or is there already an appropriate abstraction in the embedded Rust ecosystem?

datdenkikniet replied:
I would highly recommend that the [rand_core](https://crates.io/crates/rand_core) or [rand](https://crates.io/crates/rand) (which is mostly an extension to `rand_core`) crate/traits are used here! They're quite widely used.

`RngCore` is implemented for the hardware RNG in `stm32f4xx-hal`, `stm32f7xx-hal` and `stm32l4xx-hal` crates (and others?).

Dirbaio replied:
I'm not sure what to do about this. A related item is TCP initial sequence numbers should be randomized too, they're currently constant. So whatever we do about RNG for DNS, we want TCP to be able to use it too.

- Option 1: Own an instance of `T: RngCore`. This would mean adding a generic param to at least Interface, InterfaceBuilder, TcpSocket, DnsSocket, SocketSet. I really don't like dirtying the API with more generics than necessary...
- Option 2: use `dyn`. Annoying because you need some kind of `alloc` container to own a `dyn RngCore`. Borrowing a `&mut dyn RngCore` could work but forcing borrowing for std targets is annoying too. Also the vtable overhead is not nice.
- Option 3: Be opinionated and bundle a particular PRNG. Have the user set the seed. It doesn't really *have* to be cryptographically secure, so a weak small PRNG would do. It's rude to rule out using a CSPRNG though.
- Option 4: Have cargo features so the user can choose between a simple PRNG or a big fat CSPRNG. If the user is not careful this could cause bloat because they'll end up with two CSPRNGS in the binary, though.
- Option 5: Default to a weak PRNG, add a Cargo feature to delegate to a `extern "Rust" fn _smoltcp_rand() -> u32` so the user can hook up their own CSPRNG/HWRNG if wanted.

Thoughts?

sesam replied:
OEMs rarely change anything that's not possible to avoid changing, so how about an Option 6. Default to as secure as possible, but document quite clearly how to trim away security and bytes.

ryan-summers replied:
I agree that we should be "secure by default". Devs will follow the path of least resistance, and if we write it unsecure, then it's very likely most applications will be insecure out of ignorance (likely including my own). That being said, I don't see any reason we shouldn't allow an opt-out? Alternatively, we can force them to provide an `(rng: impl RngCore)` in any function when requesting a new DNS request, which would allow randomization of sockets (e.g. get random generator only when needing random numbers). This only addresses DNS though and not TCP.

Dirbaio replied:
These RNG uses are not that critical for security though. They (DNS txid, DNS source port, TCP initial sequence numbers) aim to make blind spoofing attacks harder by adding a bit more entropy (the key here is *blind*: it's an attacker who can spoof packets, but can't sniff the legit packets between the smoltcp host and the peer. If it could, it no longer offers protection: the attacker can simply see the values for port/seq/txid and spoof the right response packet).

Also, they only offer 32bits of entropy at best which is nowhere near enough to be considered secure anyway. if you care about security you should be using legit authenticated encryption such as TLS. If you are, such attacks become DoS's at best, they can't compromise you.

Someone in your LAN can DoS you in lots of different ways already, such as ARP spoofing, or flooding you at 100mbps line rate. Someone outside your LAN will have a much harder time spoofing packets in the first place: they have to get past NAT/firewall.

Also, bundling our own CSPRNG by default still doesn't make it secure-by-default: the user has to seed it securely, which is necessarily hardware-dependent. I don't see an easy way to get past that...

Dirbaio replied:
Something we can do is simply *not* bundle a weak PRNG by default, forcing the user to supply some impl. If they decide to supply a weak one, it's their problem.

Dirbaio replied:
I've opened a separate PR for the rand infra: #547 

What do you guys think? Feedback/reviews welcome!

Dirbaio committed:
wire: add DNS

Dirbaio committed:
socket: add DNS

korken89 committed:
defmt

korken89 committed:
Make Dns's QueryHandle public

korken89 committed:
dns: Added timeout and trying to use different servers

korken89 committed:
dns: Failure if no addresses are received

korken89 committed:
dns: return failure on NXDOMAIN.

korken89 committed:
DnsSocket: Be able to update servers

Dirbaio committed:
dns: use Vec for server address list.

Dirbaio committed:
dns: fix panic when servers.len = 0

Dirbaio committed:
dns: allow specifying name in human-friendly format.

Dirbaio committed:
dns: don't relaunch query on CNAMEs.

The recursive resolver is supposed to do that for us.

(note, the old code already didn't relaunch it, but the intention was
to make it do that)

Dirbaio committed:
dns: feature and unused fixes

Dirbaio committed:
dns: add waker support.

Dirbaio committed:
dns: add to CI

Dirbaio committed:
dns: "fix" wire tests on ipv6
